docs/12.EXT-REFERENCE-MODULE.md

```diff
@@ -150,13 +154,31 @@ jerry_value_t js_module = jerryx_module_resolve (requested_module, resolvers, 2)
 
 # Module helper macros
 
+## JERRYX_NATIVE_MODULE_INIT
+
+**Summary**
+
+Expands to the initializer of a variable of type `jerryx_native_module_t`.
+
+**Prototype**
+
+```c
+#define JERRYX_NATIVE_MODULE_INIT(module_name, on_resolve_cb)
+```
+
+- `module_name` - literal string giving the module's name
+- `on_resolve_cb` - the function of type `jerryx_native_module_on_resolve_t` that will be called when the module needs
+to be loaded.
+
 ## JERRYX_NATIVE_MODULE
 
 **Summary**
 
 Helper macro to define a JerryScript module. Currently stores the name of the module and its initializer in an
-executable linker section. This macro is available only if the preprocessor directive `JERRYX_NATIVE_MODULES_SUPPORTED`
-is defined.
+executable linker section.
```

Comment:
@gabrielschulhof I have a question because we may use jerryx module on linux based platform. If we run jerryscript on platform that support dynamically library loading (e.g., `dlopen`), `JERRYX_NATIVE_MODULE` macro may not store any information in an executable linker section, and it will do nothing. Do you assume the jerryx module will work on statically linked system only? Could you please correct me if I don't understand correctly?

Reply:
@glistening if you're using jerryscript in an environment that supports `dlopen()` then we should definitely add support for constructor-based loading.

Reply:
I feel this is getting out of hand. Perhaps we should move a step back and re-think what we want to do. Here is my thoughts:

a) The module resolver system with the canonical name support is good. It definitely covers the typical use case.
b) The native system should focus only the static modules. No dynamic loading.
c) We should provide a system for dynamic modules. Question: why do we need constructors? When the module is loaded, a startup routine should run. That startup can be defined by name (e.g.: module name underscore start(), or simply start() since dlsym has a .so argument).

Reply:
@zherczeg using constructors unifies the process of dynamic loading with that of static loading, and, like linker sections, makes it possible to avoid having to maintain an array with all possible modules.

I'm starting to think that perhaps we should support only constructors, and we should drop support for linker sections altogether, because if the C library or the compiler does not support constructors/destructors, then those functions can be called manually, since at that point the use of a global array is unavoidable.